compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs

```diff
@@ -583,3 +607,23 @@ pub fn report_dyn_incompatibility<'tcx>(
 
     err
 }
+
+fn attempt_dyn_to_impl_suggestion(tcx: TyCtxt<'_>, hir_id: Option<hir::HirId>, err: &mut Diag<'_>) {
+    let Some(hir_id) = hir_id else { return };
+    let hir::Node::Ty(ty) = tcx.hir_node(hir_id) else { return };
+    let hir::TyKind::TraitObject([trait_ref, ..], ..) = ty.kind else { return };
+    let mut hir_id = hir_id;
+    while let hir::Node::Ty(ty) = tcx.parent_hir_node(hir_id) {
+        hir_id = ty.hir_id;
+    }
+    if tcx.parent_hir_node(hir_id).fn_sig().is_none() {
```

Comment:
What does this actually suppress?

Reply:
Also, I just re-read the whole diff. I see that you just moved this code to generalize the suggestion, right?

If so, then if you don't want to deal w/ the nits in this specific area, I guess that's fine.

Reply:
I'm happy to make drive-by cleanups-- thanks for the suggestions and for helping me learn!

Reply:
> What does this actually suppress?

It ensures that we only suggest moving `dyn Trait` to `impl Trait` when in function signatures. [Here's what it looks like if this filter is removed.](https://github.com/rust-lang/rust/commit/98a432345a9c7ae49e93e039d95a9da8b93bf8b6)

Reply:
i've added a comment to this effect.